### DISCUSSION TITLE
Avoid query timeouts and out-of-memory errors by using automatic partitions

### DISCUSSION THREAD

**Discussion by jubingc** [milvus-io/milvus#37226] [Q&A and General discussion] [answered]
Hi,

Is the following the right way to use automatic partition? I assume the partitions are automatically created in collection creation, partitions are automatically used when inserting data and searching data if partition key is provided.

When putting all data (50M records) into the default partitions, we encountered search timeouts and out-of-memory errors. Now, we are experimenting with the automatic partition by having 20-100K records in a partition. And data nodes has enough memory. Currently, data node pods have requests memory 32 GB and limit memory 64 GB. The memory used in data node pods is under 90% of requests memory. Is there anything we should adjust or look into?

We are creating partitions like this:
```python
schema = MilvusClient.create_schema(
    auto_id=False,
    enable_dynamic_field=True,
    partition_key_field="month",
    num_partitions=100
)
```

Create index:
```
index_params.add_index(
    field_name="month",
    index_type="Trie"
)
```

Prepare data:
```
    data.append({
        "vector": [ ... ],
        "month": "202410",
       ....
    })
```

Search data:
```
client.search(
    data=[...],
    filter="month == '202410'",
    search_params={...},
   ....
)
```

**Comment by xiaofan-luan** [comment]
> Hi,
> 
> Is the following the right way to use automatic partition? I assume the partitions are automatically created in collection creation, partitions are automatically used when inserting data and searching data if partition key is provided.
> 
> When putting all data (50M records) into the default partitions, we encountered search timeouts and out-of-memory errors. Now, we are experimenting with the automatic partition by having 20-100K records in a partition. And data nodes has enough memory. Currently, data node pods have requests memory 32 GB and limit memory 64 GB. The memory used in data node pods is under 90% of requests memory. Is there anything we should adjust or look into?
> 
> We are creating partitions like this:
> 
> ```python
> schema = MilvusClient.create_schema(
>     auto_id=False,
>     enable_dynamic_field=True,
>     partition_key_field="month",
>     num_partitions=100
> )
> ```
> 
> Create index:
> 
> ```
> index_params.add_index(
>     field_name="month",
>     index_type="Trie"
> )
> ```
> 
> Prepare data:
> 
> ```
>     data.append({
>         "vector": [ ... ],
>         "month": "202410",
>        ....
>     })
> ```
> 
> Search data:
> 
> ```
> client.search(
>     data=[...],
>     filter="month == '202410'",
>     search_params={...},
>    ....
> )
> ```

right now all the data needs to be put into memory. 
50M data might not be able to fit into 32GB memory. this has nothing to do with partition I guess.

You can use month as the partitionkey, this help you to do automatic parititons. Bug gain unless u use IVFSQ, IVFPQ or diskann, you need to have enough memory to load all data into memory

**Reply by jubingc** [reply]
32 GB memory is per pod of the query node. I want to make sure 32 GB is enough for a single query node pod.

We currently have 10+ query node pods in the Milvus cluster managed by HPA. More query nodes will be added to the cluster automatically when the resources are not sufficient.

**Comment by xiaofan-luan** [comment]
then the memory is far more than enough.

the other issue is 20-100K per partitions, that means you have too many parititons. 

partition key feature is what you need

**Reply by jubingc** [reply]
I assume extra memory should not be the performance bottleneck, right?

Each record has a 1024 dimension vector and 2KB of other data, so each record consumes about 6 KB of memory (4 KB * 1024 + 2KB). 50 M records will consume 300GB of memory (6 KB * 50 M = 300 GB). Each query node has 32 GB of memory (29 GB is available if 90% of memory is allocated). Around 10 query nodes are reasonable.

**Reply by jubingc** [reply]
> partition key feature is what you need

Can you iterate more on this?

**Reply by xiaofan-luan** [reply]
https://milvus.io/docs/use-partition-key.md

**Comment by yhmo** [comment]
An example of partition keys:

```
import random
from pymilvus import (
    MilvusClient, DataType
)

dim = 1024
collection_name = "xxx"
milvus_client = MilvusClient("http://localhost:19530")

milvus_client.drop_collection(collection_name)

schema = milvus_client.create_schema()
schema.add_field(field_name="id", datatype=DataType.INT64, is_primary=True, auto_id=False)
schema.add_field(field_name="vector", datatype=DataType.FLOAT_VECTOR, dim=dim)
schema.add_field(field_name="month", datatype=DataType.VARCHAR, max_length=100, is_partition_key=True)

index_params = milvus_client.prepare_index_params()
index_params.add_index(field_name = "vector", index_type="IVF_FLAT", metric_type="L2", params = {"nlist":128})

milvus_client.create_collection(
    collection_name=collection_name,
    schema=schema,
    index_params=index_params,
    num_partitions=100,
)

batch = 1000
for i in range(10):
    rows = [
        {"id": i*batch + k, "vector": [random.random() for _ in range(dim)], "month": "2024" + str(k%12).zfill(2)} for k in range(batch)
    ]
    res = milvus_client.insert(collection_name=collection_name, data=rows)
    print("insert", i)

res = milvus_client.query(collection_name=collection_name, consistency_level="Strong", filter="", output_fields=["count(*)"])
print(res) # row count

topk = 5
result = milvus_client.search(collection_name=collection_name,
                           data=[[random.random() for _ in range(dim)]],
                           filter="month == '202410'",
                           limit=topk,
                           consistency_level="Bounded",
                           output_fields=["month"])
for hits in result:
    print(f"Top{topk} of search result:")
    for hit in hits:
        print(hit)
```

**Comment by yhmo** [comment, marked as answer]
For high-dimension vectors, the required memory mainly depends on the index type.

Typically, the index size compare to the original vector size:
- IVF_FLAT/HNSW, a bit larger than the original vector size
- IVF_SQ8/IVF_PQ, 25% ~ 30% of the original vectors size, the recall rate of IVF_PQ is not good.
- DISKANN, 1/4 ~ 1/6 of the original vectors size, performance is not good, requires high-performance hard disk.

50M 1024dim vectors, the original size is 200GB.
- For IVF_FLAT/HNSW index, the required memory size is 200GB+
- For IVF_SQ8/IVF_PQ index, the required memory size is 50GB ~ 60GB
- For DISKANN index, the required memory size is 40GB ~ 50GB

The other data is also loaded into memory. Typically, 100GB of data requires 100GB of memory.
So, theoretically:
- For IVF_FLAT/HNSW index, the total memory usage is 200GB+100GB
- For IVF_SQ8/IVF_PQ index, the total memory usage is 50GB ~ 60GB + 100GB
- For DISKANN index, the total memory usage is 40GB ~ 50GB + 100GB

If you choose IVF_FLAT/HNSW, and there are 10 query nodes with 32GB per node, it is a bit risky to load the collection. You might get "unable to load" warning because the query node has memory protection logic. If the memory usage is near the max capacity, it will reject loading.

**Reply by jubingc** [reply]
@yhmo Thanks a lot for sharing this information. It is really helpful for resource estimation.

Is there any document describing this in detail? Is it possible to have similar resource estimation for other indexing.

**Reply by yhmo** [reply]
These theoretical sizes are based on the index algorithms. In runtime, the required resources are usually much larger than the theoretical size. We don't intend to write it into document because using the theoretical size to deploy resources is not safe.
We officially have a tool to estimate the required resources by data size: https://milvus.io/tools/sizing
The tool shows a larger size than the theoretical size to avoid potential OOM.

**Reply by jubingc** [reply]
Thanks. I am curious about the memory size of query nodes. If data is loaded by segments into query nodes, and every segment is using the default size of 512 MB, is any of the following configurations preferred over others?

1. 4 core 16 GB x 24
2. 8 core 32 GB x 12
3. 16 core 64 GB x 6

**Reply by yhmo** [reply]
Just let the tool recommend resources for you. Input the dimension of your vectors, row count of your data, index type, let the tool recommend.


![Screenshot from 2024-10-31 14-24-15](https://github.com/user-attachments/assets/614fef31-a800-44db-96d5-92cc0da6a502)

**Reply by jubingc** [reply]
Yes, I understand. Let me rephrase my question.

All the configurations below have the same total CPU and memory capacity (96-core CPU and 384 GB of memory), and all should comfortably handle 512 MB segments. However, the recommended instance size varies based on data size. I'm curious about the rationale for preferring one instance size over the others. Thanks!

1. 4 core 16 GB x 24
2. 8 core 32 GB x 12
3. 16 core 64 GB x 6
